include/seqan3/alphabet/detail/alphabet_base.hpp

```diff
@@ -0,0 +1,237 @@
+
+    /*!\name Member types
+     * \{
+     */
+    //!\brief The type of the alphabet when converted to char (e.g. via to_char()).
+    using char_type = char_t;
+    //!\brief The type of the alphabet when represented as a number (e.g. via to_rank()).
+    using rank_type = detail::min_viable_uint_t<size - 1>;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_base() = default;
+    constexpr alphabet_base(alphabet_base const &) = default;
+    constexpr alphabet_base(alphabet_base &&) = default;
+    constexpr alphabet_base & operator=(alphabet_base const &) = default;
+    constexpr alphabet_base & operator=(alphabet_base &&) = default;
+    //!\}
+
+    /*!\name Read functions
+     * \{
+     */
+    /*!\brief Return the letter as a character of char_type.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::to_char() requirement via the seqan3::to_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr char_type to_char() const noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
+    //!\endcond
+    {
+        return derived_type::rank_to_char[rank];
+    }
+
+    /*!\brief Return the letter's numeric value or rank in the alphabet.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::semi_alphabet_concept::to_rank() requirement via the to_rank() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr rank_type to_rank() const noexcept
+    {
+        return rank;
+    }
+    //!\}
+
+    /*!\name Write functions
+     * \{
+     */
+    /*!\brief Assign from a character.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::assign_char() requirement via the seqan3::assign_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr derived_type & assign_char(std::conditional_t<std::Same<char_type, void>, char, char_type> const c) noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
+    //!\endcond
+    {
+        using index_t = std::make_unsigned_t<char_type>;
+        rank = derived_type::char_to_rank[static_cast<index_t>(c)];
+        return static_cast<derived_type &>(*this);
+    }
+
+    /*!\brief Assign from a numeric value.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::semi_alphabet_concept::assign_rank() requirement via the seqan3::assign_rank() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr derived_type & assign_rank(rank_type const c) noexcept
+    {
+        assert(c < value_size);
+        rank = c;
```

Comment:
what happens in release mode here if the assertion does not catch?

Reply:
Undefined Behaviour.